src/target/riscv/riscv-013.c

```diff
@@ -3764,35 +3776,51 @@ static int write_memory_bus_v1(struct target *target, target_addr_t address,
 			if (riscv_batch_available_scans(batch) < (size + 3) / 4)
 				break;
 
-			if (size > 12)
-				riscv_batch_add_dmi_write(batch, DM_SBDATA3,
-						((uint32_t) p[12]) |
+			uint64_t value128_h = 0;
+			uint64_t value128_l = 0;
+			uint32_t value32;
+			if (size > 12) {
+				value32 = ((uint32_t) p[12]) |
 						(((uint32_t) p[13]) << 8) |
 						(((uint32_t) p[14]) << 16) |
-						(((uint32_t) p[15]) << 24), false);
+						(((uint32_t) p[15]) << 24);
+				riscv_batch_add_dmi_write(batch, DM_SBDATA3, value32, false);
+				value128_h = ((uint64_t)value32 << 32);
+			}
 
-			if (size > 8)
-				riscv_batch_add_dmi_write(batch, DM_SBDATA2,
-						((uint32_t) p[8]) |
+			if (size > 8) {
+				value32 = ((uint32_t) p[8]) |
 						(((uint32_t) p[9]) << 8) |
 						(((uint32_t) p[10]) << 16) |
-						(((uint32_t) p[11]) << 24), false);
-			if (size > 4)
-				riscv_batch_add_dmi_write(batch, DM_SBDATA1,
-						((uint32_t) p[4]) |
+						(((uint32_t) p[11]) << 24);
+				riscv_batch_add_dmi_write(batch, DM_SBDATA2, value32, false);
+				value128_h = value128_h | ((uint64_t)value32);
+			}
+			if (size > 4) {
+				value32 = ((uint32_t) p[4]) |
 						(((uint32_t) p[5]) << 8) |
 						(((uint32_t) p[6]) << 16) |
-						(((uint32_t) p[7]) << 24), false);
-			uint32_t value = p[0];
+						(((uint32_t) p[7]) << 24);
+				riscv_batch_add_dmi_write(batch, DM_SBDATA1, value32, false);
+				value128_l = ((uint64_t)value32 << 32);
+			}
+
+			value32 = p[0];
 			if (size > 2) {
-				value |= ((uint32_t) p[2]) << 16;
-				value |= ((uint32_t) p[3]) << 24;
+				value32 |= ((uint32_t) p[2]) << 16;
+				value32 |= ((uint32_t) p[3]) << 24;
 			}
 			if (size > 1)
-				value |= ((uint32_t) p[1]) << 8;
-			riscv_batch_add_dmi_write(batch, DM_SBDATA0, value, false);
+				value32 |= ((uint32_t) p[1]) << 8;
+
+			value128_l = value128_l | ((uint64_t)value32);
+			riscv_batch_add_dmi_write(batch, DM_SBDATA0, value32, false);
+
+			if (size == 16)
+				log_memory_access128(address + i * size, value128_h, value128_l, false);
+			else
+				log_memory_access(address + i * size, value128_l, size, false);
```

Comment:
Thank you for this change.

For improved readability, I'd recommend to make four `uint32_t` variables `sbdata0`, `sbdata1`, `sbdata2` and `sbdata3`, initialize them to zero and then assign them in the `if`s.

The last part would then look this way:

```suggestion
			if (size == 16) {
				uint64_t value_h = (sbaccess3 << 8) + sbaccess2;
				uint64_t value_l = (sbaccess1 << 8) + sbaccess0;
				log_memory_access128(address + i * size, value_h, value_l, false);
			} else {
				uint64_t value = (sbaccess1 << 8) + sbaccess0;			
				log_memory_access(address + i * size, value128_l, size, false);
			}
```

Reply:
done